### ISSUE TITLE
Create constructors with a defaulted cache

### ISSUE THREAD

**Issue by aleokdev** [mapeditor/rs-tiled#187]
Right now all of our examples and test code uses a default `&mut FilesystemResourceCache` for loading, as it only handles a single map. It would be wise to expose a loading method that doesn't take in a cache and another one that does (with a `_cached` postfix) as it would confuse new users less and would result in less boilerplate for small applications and whatnot.

**Comment by bjorn** [comment]
Since I'm not sure about providing an easy way to load files without caching tilesets, might it help to provide a "factory object", like a `MapLoader` struct that would wrap the cache and possibly other stuff we'd add in the future?

```rs
let mut loader = tiled::Loader::new();
let map = loader.load_map_file(map_path).unwrap();
```

Or, since we allow providing a custom cache:

```rs
let mut cache = CustomResourceCache::new();
let mut loader = tiled::Loader::new_with_cache(cache);
let map = loader.load_map_file(map_path).unwrap();
```

**Comment by aleokdev** [comment]
No no, there's no need. We could just construct a default cache internally from the "non-cached" function, and pass it into the cached function.

**Comment by bjorn** [comment]
My main point was to simplify things for the user by moving the exposed loading API out of the `Map` and `Tileset` structs, into an object that can internally hold the cache, such that the user doesn't need to pass a cache around.

**Comment by aleokdev** [comment]
I don't know. The loader would need to internally keep track of an `Option<T>` which should verify that `T: ResourceCache` and it'd make things more complex from the typing point of view. IMO it's not worth the cost.

**Comment by bjorn** [comment]
> I don't know. The loader would need to internally keep track of an `Option<T>` which should verify that `T: ResourceCache` and it'd make things more complex from the typing point of view. IMO it's not worth the cost.

No need for `Option`, because there would always be a resource cache. If somebody really wants to disable caching, a non-caching implementation could be used.

> Plus, instead of juggling the cache around, you'd be passing the loader. Same story.

Not really the same, because the loader doesn't necessarily need to be passed around. You brought up this issue because of the `ResourceCache` verbosity in all our tests and examples, in none of which we'd need to pass this loader around.

I'm not really sure what the plan is, but I'm just weary of providing a convenient `Map::parse_file` that internally constructs a cache, because it would set a bad example. It also makes the approach we would want to encourage even more verbose.

**Comment by aleokdev** [comment]
> No need for Option, because there would always be a resource cache. 

> I'm just weary of providing a convenient Map::parse_file that internally constructs a cache, because it would set a bad example.

Well, don't both approaches internally construct a cache?

**Comment by aleokdev** [comment]
One thing to mention: The cache doesn't necessarily have to pose overhead if we create a "NoopResourceCache" of sorts that just does nothing on insert and get. We'd have to mention in the ResourceCache docs that insert is not actually guaranteed to insert the object though.

**Comment by bjorn** [comment]
> Well, don't both approaches internally construct a cache?

Yes, but if `Map::parse_file` does it internally it's much more likely a user might end up reconstructing a cache every time, than when they need to instantiate a `Loader` each time. In second case reusing the loader is the more natural thing to do.

> The cache doesn't necessarily have to pose overhead if we create a "NoopResourceCache" of sorts that just does nothing on insert and get. We'd have to mention in the ResourceCache docs that insert is not actually guaranteed to insert the object though.

I guess it's not a good idea, especially with support for object templates coming eventually.

**Comment by aleokdev** [comment]
Okay, I see your point now. Will consider it. There are both pros and cons to each approach. The main con of my approach is that it's not obvious that the `_cached` version should be used for extra performance. The con of the loader approach is that it is more verbose, requires an extra type and it is not straightforward that map and tileset construction is done in an external type rather than from the types themselves.

**Comment by bjorn** [comment]
We could provide a convenient `Map::parse_file_with_temp_cache` instead. :-)

I'm also a little worried about the explosion of variations, once we'd add support for the JSON format (functions to load json with or without cache, for maps, tilesets, templates...), and then functions that don't take file names directly but some reader, and maybe async variations eventually. It quickly starts to add up.

**Comment by aleokdev** [comment]
> We could provide a convenient `Map::parse_file_with_temp_cache` instead. :-)

That looks better, and will require less changes. But it is maybe a bit too verbose. How about `parse_file_uncached` along with `parse_file_cached`?

> I'm also a little worried about the explosion of variations, once we'd add support for the JSON format (functions to load json with or without cache, for maps, tilesets, templates...), and then functions that don't take file names directly but some reader, and maybe async variations eventually. It quickly starts to add up.

Regardless of the approach that we take, this will be a problem in any of them. We can figure out a solution once we see that things are getting out of hand.

**Comment by bjorn** [comment]
> That looks better, and will require less changes. But it is maybe a bit too verbose. How about `parse_file_uncached` along with `parse_file_cached`?

I think `uncached` suggest there is no cache, even though there is a temporary one. And `cached` makes the desired approach needlessly more verbose (it's already clear that a cache is provided, since it goes in as a parameter), and `Map::parse_file_cached` suggests that the map file is somehow cached, even though that is not the case.

**Comment by aleokdev** [comment]
The Loader idea does look very good now that I've thought about it carefully. The main issue that I had with it is it seemed redundant being able to just place methods in Map/Tileset instead, but it makes much more sense once we insert JSON and other formats into the picture. A TMX/TSX/JSONLoader can offload all the parsing methods from the actual datatypes into a separate module and also fix the paths issue along with it. It looks great this way.

P.S. Sorry if i seem extremely dense at times, I tend to prioritize reply speed over thought :p

**Comment by bjorn** [comment]
> A TMX/TSX/JSONLoader can offload all the parsing methods from the actual datatypes into a separate module and also fix the paths issue along with it. It looks great this way.

I'm glad you're seeing the loader approach in a more positive light. I'm unsure however, whether it makes sense to have different loaders for different formats (or even different data types, as with TMX and TSX). If it comes to the cache, at least, it is format-agnostic so we don't need a separate cache for each format.

> P.S. Sorry if i seem extremely dense at times, I tend to prioritize reply speed over thought :p

It's no problem. I just think sometimes a quick chat or even video call might be better than typing comments then. They have too much overhead for this. :-)
